### ISSUE TITLE
Question regarding compatibility with MuleSoft

### ISSUE THREAD

**Issue by mcheguini** [StefH/RamlToOpenApiConverter#8]
Hello Stef,

I would like to say thank you for this amazing framework, and your support (:
I understand you implemented “ConvertToOpenApiDocument” to convert raml to OpenAPI 
I tried to use “ConvertToOpenApiDocument” and “ConvertToFile” for following example code (it is attached for your kind review)
regarding the mentioned fact, is it possible to use following capabilities in WireMock.NET?
- Uses (can be converted to .json?)
- Traits (can be converted in .json?, can I use is: to select an specific trait i.e  "is: sometrait" in body?)
-  Is it possible to use "ExactMatcher" WildCard in raml file for post body request?
- securitySchemes (it will not appear in .json file)
- example: !include Email_Request_Example.json (I am not able to include .json file)

I was able to use the following json schema as example in raml file, but the problem is each example should be placed in separate .json file (there are a lot of .json example files).
Example: |
{
	Some json data
}
Could you please help me to solve previously mentioned inquiries specially with  "example: !include examplefile.json" 

Also I found, having nested !include will cause an error, for example

```
types:
  ParentVar: !Parent.raml
```
**in Parent.raml**
```
#%RAML 1.0
type: object
properties:
  somefield:
    type: array
    items: !include child.raml
```
In mentioned "items: !include child.raml" will cause an error

**Comment by StefH** [comment]
Can you provide the all the raml files ?

**Comment by mcheguini** [comment]
Sure thing here is the example code:
test.raml

``` yml
#%RAML 1.0
title: example raml document
description: this is a description of the raml file
protocols:
  - HTTP
version: 1.0
baseUri: http://urlthatnotexist.api/api
mediaType:
  - application/json
securitySchemes:
  basic: !include common/securityScheme.raml

uses:
  ComplexLib: datatype/library.raml

traits:
  somerequirement:
    headers:
      name:
        type: string
        required: true
      id:
        type: string
        required: true

types:
  RequestEmail: !include datatype/email_request.raml
  ResponseEmail: !include datatype/email_response.raml

/someendpoint:
  post:
    description: This is an example endpoint
    is: [somerequirement]
    body:
      application/json:
        type: RequestEmail
        example: !include examples/email_request.json
          
    responses:
      200:
        body:
          application/json:
            type: ResponseEmail
            example: !include examples/email_response.json
```

securityScheme.raml
``` yml
#%RAML 1.0 SecurityScheme
type: someAuthentication
description: this API support basic auth
```

trait.raml
``` yml
#%RAML 1.0 Trait
somerequirement:
  headers:
    name:
      type: string
      required: true
    id:
      type: string
      required: true
      
queryParameters:
  offset:
    displayName: first page
    type: integer
    required: false
    example: 1
  limit:
    description: max page number
    type: integer
    required: false
    example: 99
```

email_request.raml
``` yml
#%RAML 1.0 DataType
type: object
properties:
  subject: 
    type: string
    required: true
  body: 
    type: string
    required: false
```

email_response.raml
``` yml
#%RAML 1.0 DataType
type: object
properties:
  response: 
    type: string
    required: true
```

email_request.json
``` json
{
    "subject": "email subject",
    "body" : "email body"
}
```

email_response.json
``` json
{
    "response": "email response"
}
```

**Comment by StefH** [comment]
Hello @mcheguini,

I'll investigate (probably the `!include` is not handled correctly...)

**Comment by mcheguini** [comment]
Given the array "Person" that contains another array "client" and "employee"
Person Response
![image](https://user-images.githubusercontent.com/68036561/139580530-8feb93d4-4683-4c62-90c5-68e7f4d70fc7.png)
Client Response
![image](https://user-images.githubusercontent.com/68036561/139580566-71d8d5de-8f30-4131-b836-f21dab9eaf27.png)
Employee Response
![image](https://user-images.githubusercontent.com/68036561/139580587-43db9b53-4e8e-45b4-994a-09dd039e81f1.png)
Purchase Response
![image](https://user-images.githubusercontent.com/68036561/139580619-8ed2a03c-b754-4498-a0df-649504d47458.png)
Tipos Response
![image](https://user-images.githubusercontent.com/68036561/139580632-d0f931a4-0a92-4029-b9b8-88f310cacdb3.png)
Email Response
![image](https://user-images.githubusercontent.com/68036561/139580651-decba836-5dde-4186-afa2-acb1b5fa0346.png)
Name Response
![image](https://user-images.githubusercontent.com/68036561/139580666-b81c4605-1059-4fe5-ba40-44d9e1d9d595.png)

Inside MuleSoft, in the main raml file "test.raml", when I point to "person" end point, I will receive person array with all expected json response without any issue:
![image](https://user-images.githubusercontent.com/68036561/139580762-b3cba343-7ea3-4b2e-a011-5cb51377eaf4.png)

When I use RamlToOpenApiConverter, there is an issue with "!include": apparently nested !include is not supported.
I tried to combine all mentioned raml response into the single raml response and I named it person2 endpoint

```
#%RAML 1.0 DataType
type: object
properties:
  client: 
    type: array
    items:
      type: object
      properties:
        purchase:
          type: array
          items:
            type: object
            properties:
              itemid: 
                type: integer
                required: true
              price: 
                type: integer
                required: false
              tipos:
                type: array
                items:
                  type: object
                  properties:
                    car: 
                      type: string
                      required: true
                    moto: 
                      type: string
                      required: false
                    bike:
                      type: string
                      required: false
        date: 
          type: datetime
          required: false
        cid:
          type: integer
          required: false
  employee: 
    type: array
    items:
      type: object
      properties:
        email:
          type: array
          items:
            type: object
            properties:
              response: 
                type: string
                required: true
        id: 
          type: integer
          required: true
        name: 
          type: array
          items:
            type: object
            properties:
              firstname: 
                type: string
                required: true
              lastname: 
                type: string
                required: true
```

When I tried to process person2 response using RamlToOpenApiConverter, I recieved the following json file:
As you can see RamlToOpenApiConveter is not able to process array elements(items;) it only shows them as array
![image](https://user-images.githubusercontent.com/68036561/139580985-69792580-16aa-4f76-a5e7-29854cf6d69d.png)
Could you please help me to solve this issue, or let me know how can I process array in RamlToOpenApiConverter 
I really really appreciate it (:

I used the following code to convert raml to json
![image](https://user-images.githubusercontent.com/68036561/139581067-011a5457-18a3-48cd-92f4-2eff579adff3.png)

This is a raml main code
![image](https://user-images.githubusercontent.com/68036561/139581113-e92f32d9-2f05-4c5b-b789-b8b9a97d7270.png)


The project is attached for your review
[person.zip](https://github.com/StefH/RamlToOpenApiConverter/files/7448657/person.zip)

**Comment by mcheguini** [comment]
Hi @StefH,
Regarding array datatype
Given the following raml file:
```
#%RAML 1.0
title: Hello world
types:
  Email:
    type: object
    properties:
      subject: string
  Emails:
    type: Email[]

/testendpoint:
  get:
    responses:
      200:
        body:
          application/json:
            type: Emails
```
Apparently RamlToOpenApi converter it is not able to process array objects
After little bit research and look at json file, I found the following part is missing:
```
      ,
      "Emails": {
        "type": "array",
        "items": {"$ref" : "#/components/schemas/Emails"}
      }
```

```
{
  "openapi": "3.0.1",
  "info": {
    "title": "Hello world"
  },
  "paths": {
    "/test": {
      "get": {
        "description": "get all",
        "responses": {
          "200": {
            "description": null,
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Emails"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "Email": {
        "type": "object",
        "properties": {
          "subject": {
            "type": "string"
          },
          "body": {
            "type": "string"
          }
        }
      }
    }
  }
}
```
I believe this is a reason that RamlToOpenApiConverter is not able to process array datatype,
Could you please help me to add this capability, I appreciate it.

**Comment by StefH** [comment]
https://github.com/StefH/RamlToOpenApiConverter/pull/9

**Comment by StefH** [comment]
Hello @mcheguini, the issue with `include` has been fixed, please try new NuGet.

I still need to check the array-issue...

**Comment by StefH** [comment]
Version `0.4.0` should also solve the referenced array issue.

Can you please retest?

**Comment by mcheguini** [comment]
Hi @StefH
Thank you for your support (:
- Regarding "!include", given the following raml code, RamlToOpenApiConverter is not able to process email.json and include the example file:
```
#%RAML 1.0
title: this is email include example
description: this is email description
mediaType:
  - application/json
version: "v1.0"
types:
  Email:
    type: object
    properties:
      subject: string
      body: string
  Emails:
    type: Email[]

/testemail:
  get:
    responses:
      200:
        body:
          application/json:
            type: Emails
            example: !include examples/email.json 
```

email.json:
```
[{
  "subject": "Email subject1",
  "body": "Email body1"
},
{
  "subject": "Email subject2",
  "body": "Email body2"
}]
```
Include Example files:
[ramldummy.zip](https://github.com/StefH/RamlToOpenApiConverter/files/7470244/ramldummy.zip)

- Regarding the Array, if the raml file contains the array object like the previous example it will work without an issue, but if the datatype has been defined as "type" : "array", RamlToApiConverter is not able to process and map the array contains.
Given the following raml file:

```
#%RAML 1.0
title: include array example
description: this is a description of array example file
protocols:
  - HTTP
version: 1.0
mediaType:
  - application/json

types:
  Email:
    type: object
    properties:
      subject:
        type: string
      body:
        type: string
  Emails:
    type: array
    items: Email
    uniqueItems: true

/testemails:
  get:
    description: this is an example array response
    responses:
      200:
        body:
          application/json:
            type: Emails
```

After looking at json file, following line still is missing
```
,
      "Emails": {
        "type": "array",
        "items": {
          "$ref": "#/components/schemas/Email"
        }
      }
```

OpenApi Json file
![image](https://user-images.githubusercontent.com/68036561/140189054-a44c0cfc-a5e2-4221-b1d9-2f82f7be05c7.png)

Could you please incorporate mentioned capabilities to RamlToOpenApiConverter, I really appreciate it.

**Comment by StefH** [comment]
For the array, please test 0.4.1

For the example.json : I need to check if I process examples at all...

**Comment by mcheguini** [comment]
Hi @StefH
Following issue catches my eye, given the following raml code, (defining datatype Emails that contains Email object)

```
#%RAML 1.0
title: email include
description: email include test
mediaType:
  - application/json
version: "v1"
protocols:
  - HTTP  

types:
  Emails: !include datatypes/Emails.raml
  Email: !include datatypes/Email.raml

/testemail:
  get:
    responses:
      200:
        body:
          application/json:
            type: Emails
```

Email.raml
```
#%RAML 1.0 DataType
description: "this is email object" 
type: object
properties:
  subject: string
  body: string
```

Emails.raml
```
#%RAML 1.0 DataType
type: object
description: Emails
properties:
  email?: 
    type: Email
```

looking at the json file RamlToOpenAPiConveter is not able to create the Email items inside the Emails
In following code
```
{
  "openapi": "3.0.1",
  "info": {
    "title": "email include",
    "description": "email include test",
    "version": "v1"
  },
  "paths": {
    "/testemail": {
      "get": {
        "responses": {
          "200": {
            "description": null,
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Emails"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "Emails": {
        "type": "object",
        "properties": {
          "email": {
            "type": "object"
          }
        }
      },
      "Email": {
        "type": "object",
        "properties": {
          "subject": {
            "type": "string"
          },
          "body": {
            "type": "string"
          }
        }
      }
    }
  }
}
```
Email object inside Emails doesn’t contain subject and body properties and shows just as an object.

**Comment by mcheguini** [comment]
> For the example.json : I need to check if I process examples at all...



> For the array, please test 0.4.1
> 
> For the example.json : I need to check if I process examples at all...

I tested the array capability and it works (: can you please add this capability to the next release.
I appreciate it

**Comment by StefH** [comment]
Array fix is already released as version `0.4.1`.

**Comment by mcheguini** [comment]
> 
> 
> Array fix is already released as version `0.4.1`.

**Comment by mcheguini** [comment]
> 
> 
> Hi @StefH Following issue catches my eye, given the following raml code, (defining datatype Emails that contains Email object)
> 
> ```
> #%RAML 1.0
> title: email include
> description: email include test
> mediaType:
>   - application/json
> version: "v1"
> protocols:
>   - HTTP  
> 
> types:
>   Emails: !include datatypes/Emails.raml
>   Email: !include datatypes/Email.raml
> 
> /testemail:
>   get:
>     responses:
>       200:
>         body:
>           application/json:
>             type: Emails
> ```
> 
> Email.raml
> 
> ```
> #%RAML 1.0 DataType
> description: "this is email object" 
> type: object
> properties:
>   subject: string
>   body: string
> ```
> 
> Emails.raml
> 
> ```
> #%RAML 1.0 DataType
> type: object
> description: Emails
> properties:
>   email?: 
>     type: Email
> ```
> 
> looking at the json file RamlToOpenAPiConveter is not able to create the Email items inside the Emails In following code
> 
> ```
> {
>   "openapi": "3.0.1",
>   "info": {
>     "title": "email include",
>     "description": "email include test",
>     "version": "v1"
>   },
>   "paths": {
>     "/testemail": {
>       "get": {
>         "responses": {
>           "200": {
>             "description": null,
>             "content": {
>               "application/json": {
>                 "schema": {
>                   "$ref": "#/components/schemas/Emails"
>                 }
>               }
>             }
>           }
>         }
>       }
>     }
>   },
>   "components": {
>     "schemas": {
>       "Emails": {
>         "type": "object",
>         "properties": {
>           "email": {
>             "type": "object"
>           }
>         }
>       },
>       "Email": {
>         "type": "object",
>         "properties": {
>           "subject": {
>             "type": "string"
>           },
>           "body": {
>             "type": "string"
>           }
>         }
>       }
>     }
>   }
> }
> ```
> 
> Email object inside Emails doesn’t contain subject and body properties and shows just as an object.

@StefH, could you please look at this issue (: I really appreciate it.
